src/LockTable.cc

```diff
@@ -0,0 +1,225 @@
+/* Copyright (c) 2014-2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "LockTable.h"
+#include "BitOps.h"
+#include "Memory.h"
+
+namespace RAMCloud {
+
+/**
+ * Constructor for LockTable.
+ *
+ * \param numBuckets
+ *      Locks in a LockTable are stored in one or more buckets.  This parameter
+ *      sets the number of buckets that should be used for this new LockTable.
+ *      This value should be a power of two.
+ *
+ * \throw Exception
+ *      An exception is thrown if the resulting number of buckets is 0.
+ */
+LockTable::LockTable(uint64_t numBuckets)
+    : bucketIndexHashMask(BitOps::powerOfTwoLessOrEqual(numBuckets) - 1)
+    , buckets()
+{
+    if (numBuckets != (bucketIndexHashMask + 1)) {
+        RAMCLOUD_LOG(DEBUG,
+                     "LockTable truncated to %lu buckets "
+                     "(nearest power of two)",
+                     (bucketIndexHashMask + 1));
+    }
+
+    if ((bucketIndexHashMask + 1) == 0)
+        throw Exception(HERE, "LockTable numBuckets == 0?!");
```

Comment:
Don't throw an exception here. This is a bug, right? In that case you should just assert, or call DIE.

Better yet, just define the error out of existence: if the number of buckets is too small, round it up to a minimum acceptable size.